internal/witnessdistro/witnessdistro.go

```diff
@@ -0,0 +1,154 @@
+package witnessdistro
+
+import (
+	"crypto"
+	"errors"
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/in-toto/go-witness/cryptoutil"
+	"gopkg.in/yaml.v3"
+)
+
+// Manifest represents the available versions of Witness
+type Manifest struct {
+	Versions map[string]Version `json:"versions"`
+}
+
+// Version represents a version of witness (ex 0.2.0) with the available Distributions of version
+type Version struct {
+	Distributions map[string]Distribution `json:"distributions"`
+	Description   string                  `json:"description"`
+}
+
+// Distribution is a specific distribution of Witness (ex linux-amd64)
+type Distribution struct {
+	FileLocation string `json:"-"`
+	SHA256Digest string `json:"sha256digest"`
+}
+
+type WitnessDistroStore struct {
+	manifest Manifest
+}
+
+type Option func(*WitnessDistroStore) error
+
+// WithManifest creates a WitnessDistroStore with the provided manifest
+func WithManifest(manifest Manifest) Option {
+	return func(wds *WitnessDistroStore) error {
+		wds.manifest = manifest
+		return nil
+	}
+}
+
+// WithManifestFile creates a WitnessDistroStore with a manifest loaded from a yaml file on disk
+func WithManifestFile(manifestPath string) Option {
+	return func(wds *WitnessDistroStore) error {
+		manifestBytes, err := os.ReadFile(manifestPath)
+		if err != nil {
+			return err
+		}
+
+		manifest := Manifest{}
+		if err := yaml.Unmarshal(manifestBytes, &manifest); err != nil {
+			return err
+		}
+
+		WithManifest(manifest)(wds)
+		return nil
+	}
+}
+
+// New creates a new WitnessDistroStore with the provided options
+func New(opts ...Option) (WitnessDistroStore, error) {
+	wds := WitnessDistroStore{}
+	errs := make([]error, 0)
+	for _, opt := range opts {
+		if err := opt(&wds); err != nil {
+			errs = append(errs, err)
+		}
+	}
+
+	if len(errs) > 0 {
+		return WitnessDistroStore{}, errors.Join(errs...)
+	}
+
+	if err := verifyManifest(wds); err != nil {
+		return WitnessDistroStore{}, err
+	}
+
+	return wds, nil
+}
+
+// verifyManifest ensures that each file exists on disk and that the sha256sum of the
+// files on disk match those of the manifest
+func verifyManifest(wds WitnessDistroStore) error {
```

Comment:
Other than the obvious desire to have the binaries distributed with TUF metadata, I wonder if there could be value of an added a check against the go binary metadata?

For example:
```
go version -m /usr/local/bin/witness | grep build
	build	-buildmode=exe
	build	-compiler=gc
	build	-trimpath=true
	build	CGO_ENABLED=0
	build	GOARCH=arm64
	build	GOOS=darwin
	build	vcs=git
	build	vcs.revision=aa35c1f4b1d41c87e139c2d333f09117fd0daf4f
	build	vcs.time=2023-08-10T14:25:28Z
	build	vcs.modified=true
```

Reply:
Yeah, I think in the future we probably want to expand the checks we can do here